src/modals/storage-connections/StorageConnectionsDataProvider.js

```diff
@@ -0,0 +1,121 @@
+}
+
+const attachConnection = async (connectionId, domainId) => {
+  return await enginePost(
+    `api/storagedomains/${domainId}/storageconnections`,
+    JSON.stringify({ id: connectionId })
+  )
+}
+
+const detachConnection = async (connectionId, domainId) => {
+  return await engineDelete(`api/storagedomains/${domainId}/storageconnections/${connectionId}`)
+}
+
+const fetchData = async (stgDomain) => {
+  if (!stgDomain?.id || !stgDomain?.dataCenterId) {
+    throw new Error('StorageConnectionsDataProvider: invalid Storage Domain')
+  }
+  const [allConnectionsJson, storageDomain, domainConnectionsJson] = await Promise.all([
+    engineGet('api/storageconnections'),
+    engineGet(`api/datacenters/${stgDomain.dataCenterId}/storagedomains/${stgDomain.id}`),
+    engineGet(`api/storagedomains/${stgDomain.id}/storageconnections`),
+  ])
+
+  if (!storageDomain) {
+    throw new Error('StorageConnectionsDataProvider: failed to fetch storage domain')
+  }
+
+  const allConnections = allConnectionsJson?.storage_connection
+  const domainConnections = domainConnectionsJson?.storage_connection
+
+  if (!allConnections || allConnections.error) {
+    throw new Error('StorageConnectionsDataProvider: failed to fetch storage connections')
+  }
+
+  if (!domainConnections || domainConnections.error) {
+    throw new Error('StorageConnectionsDataProvider: failed to fetch storage connections' +
+      'for storage domain ' + stgDomain.id)
+  }
+
+  const domainConnectionsIds = new Set(domainConnections.map(connection => connection.id))
+  const isConnectionAttachedToDomain = (conn) => { return domainConnectionsIds.has(conn.id) }
+  const isConnectionNotAttachedToDomain = (conn) => { return !isConnectionAttachedToDomain(conn) }
+  const allConnectionsByType = allConnections
+    .filter(connection => connection.type === storageDomain.storage?.type)
+    .map((conn) => {
+      return { ...conn, isAttachedToDomain: isConnectionAttachedToDomain(conn) }
+    })
+  const domainConnectionsByType = allConnectionsByType.filter(isConnectionAttachedToDomain)
+  const allConnectionsByTypeArranged = domainConnectionsByType
+    .concat(allConnectionsByType.filter(isConnectionNotAttachedToDomain))
+
+  return {
+    storageDomain: storageDomain,
+    domainConnections: domainConnectionsByType,
+    allConnections: allConnectionsByTypeArranged,
+  }
+}
+
+const StorageConnectionsDataProvider = ({ children, storageDomain }) => (
+  <DataProvider fetchData={() => fetchData(storageDomain)}>
+    {({ data, fetchError, fetchInProgress, fetchAndUpdateData }) => {
+      // expecting single child component
+      const child = React.Children.only(children)
+
+      // handle data loading and error scenarios
+      if (fetchError) {
+        getPluginApi().showToast(webadminToastTypes.danger, msg.storageConnectionsDataError())
+        return null
+      }
+
+      if (fetchInProgress || !data) {
+        return React.cloneElement(child, { isLoading: true })
+      }
```

Comment:
This will activate the spinner any time the `fetchAndUpdateData` is in progress (via `doRefreshConnections()` below.  To change it to only activate on the first fetch, do:
```suggestion
      if (!data) {
        return React.cloneElement(child, { isLoading: true })
      }
```